### PR TITLE
Password Reset

### PR DESCRIPTION
**Pull request by Ommy**
Backend and frontend for password reset using redis.

### CODE REVIEW

**Comment by Trinovantes** [comment]
Closes #44

File: src-backbone/app/js/controllers/authController.js [hunk 1]
```diff
@@ -1,9 +1,11 @@
-let App            = require('utils/sanaAppInstance');
-let Helpers        = require('utils/helpers');
-let AuthLayoutView = require('views/auth/authLayoutView');
-let SignupView     = require('views/auth/signupView');
-let LoginView      = require('views/auth/loginView');
-let SettingsView   = require('views/auth/settingsView');
+let App                       = require('utils/sanaAppInstance');
```
Comment (Trinovantes) [inline src-backbone/app/js/controllers/authController.js]: use `const` for importing (I didn't want to touch every file when I made the convention switch)

File: src-backbone/app/js/templates/auth/resetPasswordCompleteView.hbs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+    <div class="form-group">
+        <label for="password_confirmation" class="sr-only">{{ i18n "Confirm Password" }}</label>
+        <input type="password" name="password_confirmation" class="form-control" id="password_confirmation" placeholder="Confirm Password">
+    </div>
+
+    <div>
+        <button type="submit" class="btn btn-primary btn-block" id="submit-btn">{{ i18n "Change Password" }}</button>
+    </div>
+</form>
```
Comment (Trinovantes) [inline src-backbone/app/js/templates/auth/resetPasswordCompleteView.hbs]: add new line

File: src-backbone/app/js/utils/helpers.js [hunk 3]
```diff
@@ -85,4 +85,11 @@ module.exports = {
         element.click();
         document.body.removeChild(element);
     },
+
+    getURLParam: function(url, param) {
+        param = param.replace(/[\[]/,"\\\[").replace(/[\]]/,"\\\]");
+        var regex = new RegExp("[\\?&]"+param+"=([^&#]*)");
```
Comment (Trinovantes) [inline src-backbone/app/js/utils/helpers.js]: use let

File: src-backbone/app/js/views/auth/resetPasswordCompleteView.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+let App = require('utils/sanaAppInstance');
```
Comment (Trinovantes) [inline src-backbone/app/js/views/auth/resetPasswordCompleteView.js]: const

File: src-backbone/app/js/controllers/authController.js (hunk 1 above)
Comment (Ommy) [inline src-backbone/app/js/controllers/authController.js]: done

File: src-backbone/app/js/templates/auth/resetPasswordCompleteView.hbs (hunk 2 above)
Comment (Ommy) [inline src-backbone/app/js/templates/auth/resetPasswordCompleteView.hbs]: done

File: src-backbone/app/js/utils/helpers.js (hunk 3 above)
Comment (Ommy) [inline src-backbone/app/js/utils/helpers.js]: done, also changed url to be optional

File: src-backbone/app/js/views/auth/resetPasswordCompleteView.js (hunk 4 above)
Comment (Ommy) [inline src-backbone/app/js/views/auth/resetPasswordCompleteView.js]: done

File: src-backbone/app/js/views/auth/resetPasswordView.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+module.exports = Marionette.ItemView.extend({
+    template: require('templates/auth/resetPasswordView'),
+
+    ui: {
+        form: 'form',
+    },
+
+    events: {
+        'submit': 'onSubmit',
```
Comment (Kapin) [inline src-backbone/app/js/views/auth/resetPasswordView.js]: nit: this should be form:submit, correct? @Trinovantes

File: src-django/api/views.py [hunk 6]
```diff
@@ -83,6 +86,103 @@ def update_details(self, request):
...
+
+        try:
+            user = User.objects.get(email=body['email'])
+        except:
+            return self.error_response(
+                status.HTTP_404_NOT_FOUND,
+                'No one by that email exists')
+
+        token = PasswordResetTokenGenerator().make_token(user)
+        url_base = "https://sanaprotocolbuilder.me/"
```
Comment (Kapin) [inline src-django/api/views.py]: This should be from a config file.

Comment (Ommy) [inline src-django/api/views.py]: So @connorcimowsky created #325, so I'll wait on that work before moving it out

File: src-backbone/app/js/views/auth/resetPasswordView.js (hunk 5 above)
Comment (Trinovantes) [inline src-backbone/app/js/views/auth/resetPasswordView.js]: I'm more surprised that this event even triggered. I did not know the form event can propagate up the DOM tree. 

Should change it to `@ui.form submit`

**Comment by Trinovantes** [comment]
+1

File: src-backbone/app/js/views/auth/resetPasswordView.js (hunk 5 above)
Comment (Ommy) [inline src-backbone/app/js/views/auth/resetPasswordView.js]: Done

**Comment by Kapin** [comment]
+1

File: src-backbone/app/js/views/auth/resetPasswordCompleteView.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+const App = require('utils/sanaAppInstance');
+const Helpers = require('utils/helpers');
+
+module.exports = Marionette.ItemView.extend({
+
+    initialize: function(reset_token) {
```
Comment (connorcimowsky) [inline src-backbone/app/js/views/auth/resetPasswordCompleteView.js]: nit: I think we usually put `initialize` after the other definitions (in this case, `template`, `ui`, and `events`)?

File: src-django/api/templates/password_reset_template [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Hello,
```
Comment (connorcimowsky) [inline src-django/api/templates/password_reset_template]: Should this file have an extension?

Comment (Ommy) [inline src-django/api/templates/password_reset_template]: I mean we can always make it a `.txt` but it doesn't make a difference.

File: src-backbone/app/js/views/auth/resetPasswordCompleteView.js (hunk 7 above)
Comment (Ommy) [inline src-backbone/app/js/views/auth/resetPasswordCompleteView.js]: There is no real convention but `initialize` at the beginning makes more sense. It is also how objects are defined on Backbone's site.

Comment (connorcimowsky) [inline src-backbone/app/js/views/auth/resetPasswordCompleteView.js]: k :thumbsup:

File: src-django/api/templates/password_reset_template (hunk 8 above)
Comment (connorcimowsky) [inline src-django/api/templates/password_reset_template]: k :thumbsup:

**Comment by connorcimowsky** [comment]
lgtm

**Pull request merged**